mmgen/models/diffusions/utils.py

```diff
@@ -0,0 +1,245 @@
+import torch
+
+
+def _get_noise_batch(noise,
+                     image_shape,
+                     num_timesteps=0,
+                     num_batches=0,
+                     timesteps_noise=False):
+    """Get noise batch. Support get sequeue of noise along timesteps.
+    If passed noise is a torch.Tensor,
+        1. timesteps_noise == True
+            1.1. dimension of noise is 4
+                1.1.1. noise shape as [bz, c, h, w]
+                    --> expand noise to [n, bz, c, h, w]
+                1.1.2. noise shape as [n, c, h, w]
+                    --> expand noise to [n, bz, c, h, w]
+                1.1.3. noise shape as [n*bz, c, h, w]
+                    --> view noise to [n, bz, c, h, w]
+            1.2 dimension of noise is 5
+                --> return noise
+        2. timesteps_noise == False
+            2.1 dimension of noise is 3 --> view to [1, c, h, w]
+            2.2 dimension of noise is 4 --> do nothing
+    If passed noise is a callable function or None
+        1. timesteps_noise == True
+            --> generate noise shape as [n, bz, c, h, w]
+        2. timesteps_noise == False
+            --> generate noise shape as [bz, c, h, w]
+    To be noted that, we do not move the generated noise to target device in
+    this function because we can not get which device the noise should move to.
+
+    Args:
+        noise (torch.Tensor | callable | None): You can directly give a
+            batch of noise through a ``torch.Tensor`` or offer a callable
+            function to sample a batch of noise data. Otherwise, the
+            ``None`` indicates to use the default noise sampler.
+        image_shape (torch.Size): Size of images in the diffusion process.
+        num_timesteps (int, optional): Total timestpes of the diffusion and
+            denoising process. Defaults to 0.
+        num_batches (int, optional): The number of batch size. To be noted that
+            this argument only work when the input ``noise`` is callable or
+            ``None``. Defaults to 0.
+        timesteps_noise (bool, optional): If True, returned noise will shape
+            as [n, bz, c, h, w], otherwise shape as [bz, c, h, w].
+            Defaults to False.
+        device (str, optional): If not ``None``, move the generated noise to
+            corresponding device.
+    Returns:
+        torch.Tensor: Generated noise with desired shape.
+
+    """
+    if isinstance(noise, torch.Tensor):
+        # conduct sanity check for the last three dimension
+        assert noise.shape[-3:] == image_shape
```

Comment:
By the way, I found an elegant [way](https://github.com/NVlabs/stylegan3/blob/a5a69f58294509598714d1e88c9646c3d7c6ec94/torch_utils/misc.py#L82) of asserting tensor shape. Just a mention, I will submit a new PR to support this.

Reply:
StyleGAN3's `assert_shape` seems designed for exporting script models. We need more motivation to use that way.